### PR TITLE
Address last suggestion in PR #33973

### LINKED ISSUE #34303: Address last suggestion in PR #33973
**Description:**
$subject. PR #33973

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes #34303 

## Approach
N/A

## Samples
N/A

## Remarks
N/A

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -1189,36 +1189,30 @@ private BField createFieldWithType(BField field, List<BType> bTypes) {
...
-     * Get the resultant union or non-union type from a {@code List<BType>}.
+     * Get the resultant type from a {@code List<BType>}.
      *
      * @param bTypes bType list (size > 0)
      * @return {@code BUnionType} if effective members in list is > 1. {@code BType} Otherwise.
      */
-    private BType getResultantUnionOrNonUnionType(List<BType> bTypes) {
-        if (bTypes.size() == 1) {
-            return bTypes.get(0);
-        }
```
Comment (lochana-chathura) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1200]: We had missed getting the reference type for size == 1 scenario. A test case was also added for this in this pr.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -1189,36 +1189,30 @@ private BField createFieldWithType(BField field, List<BType> bTypes) {
...
-
+    private BType getResultantType(List<BType> bTypes) {
         LinkedHashSet<BType> bTypeSet = new LinkedHashSet<>(bTypes);
-        List<BType> eBTypes = new ArrayList<>(bTypes.size());
-        addEffectiveMemberTypes(eBTypes, bTypeSet);
+        List<BType> flattenBTypes = new ArrayList<>(bTypes.size());
+        addFlattenMemberTypes(flattenBTypes, bTypeSet);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1200]:
```suggestion
        List<BType> flattenedTypes = new ArrayList<>(bTypes.size());
        addFlattenedMemberTypes(flattenedTypes, bTypeSet);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -1189,36 +1189,30 @@ private BField createFieldWithType(BField field, List<BType> bTypes) {
...
 
-        return getRepresentativeBroadType(eBTypes);
+        return getRepresentativeBroadType(flattenBTypes);
     }
 
-    private void addEffectiveMemberTypes(List<BType> eBTypes, LinkedHashSet<BType> bTypes) {
+    private void addFlattenMemberTypes(List<BType> flattenBTypes, LinkedHashSet<BType> bTypes) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1205]: Isn't this now the same as `org.wso2.ballerinalang.compiler.semantics.analyzer.Types#getAllTypes`?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/record-constraint-table-value.bal [hunk 4]
```diff
@@ -658,6 +659,20 @@ function testTableTypeInferenceWithVarType10() {
...
+type NaturalNums 1|2|3;
+
+type WholeNums Zero|NaturalNums;
+
+function testTableTypeInferenceWithVarType11() {
+    WholeNums f = 2;
+
+    var v1 = table [{a: f}];
+    table<record {|(0|1|2|3) a;|}> _ = v1;
+    v1.add({a: f});
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/table/record-constraint-table-value.bal:673]: We should assert something here, right?

**Review by MaryamZi — COMMENTED** [review]
Let's change the base to https://github.com/ballerina-platform/ballerina-lang/tree/ballerina-2.0.x.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/record-constraint-table-value.bal (hunk 4 above)
Comment (lochana-chathura) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/table/record-constraint-table-value.bal:696]: Added with 6dd03ab

**Comment by lochana-chathura** [comment]
Closing Temporary.

**Review by gimantha — APPROVED** [review]

**Pull request merged**